City-of-Helsinki/atv
# ATV-56 | Setup object permissions

## Description :sparkles:
* Add `Service` object permissions with [`django-guardian`](https://django-guardian.readthedocs.io/en/stable/index.html)
* Add permission checking for API
* Setup some custom exceptions and ignore them for Sentry reporting

## Issues :bug:
### Closes :no_good_woman:
**[ATV-56](https://helsinkisolutionoffice.atlassian.net/browse/ATV-56):** Set up object permissions for groups

## Testing :alembic:
### Automated tests :gear:️
```shell
$ pytest documents/tests/test_api_list_documents.py
```

### Manual testing :construction_worker_man:
1. Install the deps (`pip-sync requirements.txt requirements-dev.txt`) and run the migrations
2. Create a new service `Test Service`, you should see the new `Group` as well from Django admin
3. Create a new user `test-service-user` from Django admin and assign it the new group `Test Service`
4. Create another service `Test Service 2`
5. Create a few documents for each service
6. Login to the Django admin with the `test-service-user`
7. Go to the documents endpoint on the API (http://localhost:8000/v1/documents/), you should see _only_ the documents associated to the `Test Service`
8. Login to the Django admin now as the superuser/admin
9. Go to the documents endpoint again, you should see all available documents

## Additional notes :spiral_notepad:
There's a few things to note or consider:

Until we implement the JWT authentication, we fetch the service based on the logged user groups (see `atv/middlewares.py`). To be able to interact with a Service's resources, a user must belong to the `Group` associated with that `Service`. Superusers (as defined by Django) are excluded from this checks and will always get full permissions to operate on all available resources, useful especially for testing/debugging.

When a new Service is created or deleted, there are corresponding signals to make sure that the corresponding permission group is created and the permissions for that service's documents are assigned to it.

The decorator's logic was taken almost entirely from [Open City Profile's implementation](https://github.com/City-of-Helsinki/open-city-profile/blob/281e28d2978a9dfe940601660f6c420e4c1821a2/open_city_profile/decorators.py).

The permission checking from the API has been moved from class-based (DRF permission classes) to decorators per method. This allows for finer grain control of the permissions for each specific action.
